### PR TITLE
feat: basic authentication with password

### PR DESCRIPTION
**Pull request by Cobidex**
# What does this PR do?
This PR implements an endpoint for authenticating users by their passwords and email. It includes:

Added POST /api/v1/auth/login Endpoint: Allows clients to login to their accounts.
Response Handling: Provides appropriate responses for missing or invalid refresh tokens.
Error Handling: Returns specific HTTP status codes and messages based on successful login
#How should this be manually tested?
Register a User:

Use the POST /api/v1/auth/register endpoint to register a new user.

Make a POST request to /api/v1/auth/login with the registered users password and email in the request body
Check the response for appropriate status codes and messages.
# Edge Cases:

Test with an invalid email to ensure the endpoint returns a 401 Unauthorized status with the correct message.
Test with an invalid password to ensure the endpoint returns a 401 Unauthorized status with the correct message

# Checklist
 Implemented POST /api/v1/auth/login endpoint in the auth.controller.ts.
 Added login method in the AuthenticationService.
 Modified auth.controller.ts to provide appropriate responses for various cases.
 Tested endpoint functionality to ensure it correctly handles valid and invalid tokens.
 Added appropriate HTTP status codes and response messages for edge cases.
 Added test for this endpoint

# Responses:
## 200 OK(JSON):
```
{
  message: "login successful"
  data: {
    user: {
      "first_name": <string>
      "last_name": <string>
      "email": <string>
      "id": <string>
    },
  },
  "status_code": 200,
  "access_token": <string>
}
```
## Error(Json):
```
{
  "message": "Invalid email or password",
  "error": "Unauthorized"
  "status_code": 401,
}
```

link to this issue: https://github.com/hngprojects/hng_boilerplate_nestjs/pull/10

N/B: There is a seperate endpoint for refresh token generation: https://github.com/hngprojects/hng_boilerplate_nestjs/pull/356

### CODE REVIEW

**Review by Heba-WebDev — APPROVED** [review]

**Comment by Cobidex** [comment]
Hello mentor @buka4rill , I have gotten approval from @Heba-WebDev , I have resolved the conflicts like you requested in the previous pr. Please review

**Review by TheCodeGhinux — APPROVED** [review]

**Comment by Cobidex** [comment]
@buka4rill waiting on your review. I have got approval to merge

File: src/modules/auth/auth.service.ts
Comment (nedssoft) [inline src/modules/auth/auth.service.ts:112]: Are you encoding user's identifiable information in the jwt token?

**Review by nedssoft — CHANGES_REQUESTED** [review]
Good job you've done! However, below are few things I think need improvement:

- There's no end-to-end test fro the login endpoint
- There's too much information in the jwt token. JWT can be decoded by anyone who have access to it, hence encoding PII like email is a security flaw. userId should be enough payload for the JWT
- It doesn't look like the JWT expiry was included in the encoding?

**Comment by Cobidex** [comment]
@nedssoft review noted, I will modify and revert

**Review by nedssoft — APPROVED** [review]
LGTM

**Pull request merged**
